docs/pages/experiments/base/popup.tsx

```diff
@@ -24,32 +25,35 @@ const Section = styled('div')`
   width: 150px;
 `;
 
-function Animated(
-  props: React.PropsWithChildren<{
-    className?: string;
-    requestOpen: boolean;
-    onExited: () => void;
-    onEnter: () => void;
-  }>,
-) {
-  const { requestOpen, onEnter, onExited, children, className } = props;
+function Animated(props: React.PropsWithChildren<{ className?: string }>) {
+  const { className, children } = props;
+  const transitionContext = React.useContext(TransitionContext);
+  if (!transitionContext) {
+    throw new Error('Missing transition context');
+  }
 
-  const handleAnimationEnd = React.useCallback(() => {
-    if (!requestOpen) {
-      onExited();
-    }
-  }, [onExited, requestOpen]);
+  const { requestedEnter, onEntering, onEntered, onExiting, onExited, hasExited } = useTransition();
 
   React.useEffect(() => {
-    if (requestOpen) {
-      onEnter();
+    if (requestedEnter && !hasExited) {
+      onEntering();
+    } else if (!requestedEnter && hasExited) {
+      onExiting();
     }
-  }, [onEnter, requestOpen]);
+  }, [onEntering, onExiting, requestedEnter, hasExited]);
+
+  const handleAnimationEnd = React.useCallback(() => {
+    if (!requestedEnter) {
+      onExited();
+    } else {
+      onEntered();
+    }
+  }, [onExited, onEntered, requestedEnter]);
```

Comment:
It's up to developer to notify when a transition ends because we don't know how to detect it (it could be a CSS transition, CSS animation, JS-based animation, etc.; they all have different methods of figuring out when it finishes).